### PR TITLE
Re-format prettyprinter-configurable with Fourmolu

### PR DESCRIPTION
**Pull request by Unisay**
- [x] Apply fourmolu to the `prettyprinter-configurable` package.
- [x] Apply `stylish-haskell` on top of it.

Closes https://github.com/IntersectMBO/plutus/issues/6682

### CODE REVIEW

File: prettyprinter-configurable/src/Text/Fixity/Internal.hs [hunk 1]
```diff
@@ -205,74 +206,85 @@ Implementing all of that is left as future work.
...
+The need for enclosing is determined from an outer 'RenderContext' and the inner fixity.
+-}
 encloseIn
-    :: Ord prec
-    => (a -> a)                -- ^ Enclose a value of type @a@ in parens.
-    -> RenderContextOver prec  -- ^ An outer context.
-    -> FixityOver prec         -- ^ An inner fixity.
-    -> a
-    -> a
+  :: (Ord prec)
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/Fixity/Internal.hs:265]: Why didn't it move `::` to the previous line?

File: prettyprinter-configurable/src/Text/Fixity/Internal.hs [hunk 2]
```diff
@@ -205,74 +206,85 @@ Implementing all of that is left as future work.
...
-            (LeftAssociative, ToTheLeft)   -> id      -- No need for parens in @Add (Add x y) z@
-                                                      -- which is rendered as @x + y + z@.
-            (RightAssociative, ToTheRight) -> id      -- No need for parens in @Concat xs (Concat ys zs)@
-                                                      -- which is rendered as @xs ++ ys ++ zs@.
-            _                              -> parens  -- Every other case requires parens.
+  case precOut `compare` precInn of
+    LT -> id -- If the outer precedence is lower than the inner, then
+    -- do not add parens. E.g. in @Add x (Mul y z)@ the precedence
+    -- of @Add@ is lower than the one of @Mul@, hence there is
+    -- no need for parens in @x + y * z@.
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/Fixity/Internal.hs:279]: It was nice and readable and now it's just utterly disgusting. Could you please return the old version (modulo indentation) and turn off `fourmolu` in this file? Oh, and please remove the stupid newline from the comment above.

File: prettyprinter-configurable/src/Text/Fixity/Internal.hs [hunk 3]
```diff
@@ -205,74 +206,85 @@ Implementing all of that is left as future work.
...
+    GT -> parens -- If the outer precedence is greater than the inner, then
+    -- do add parens. E.g. in @Mul x (Add y z)@ the precedence
+    -- of @Mul@ is greater than the one of @Add@, hence
+    -- parens are needed in @x * (y + z)@.
+    EQ -> case (assocOut, dir) of -- If precedences are equal, then judge from associativity.
+      _ | assocOut /= assocInn       -> parens -- Associativities differ => parens are needed.
+      (LeftAssociative, ToTheLeft)   -> id -- No need for parens in @Add (Add x y) z@
+      -- which is rendered as @x + y + z@.
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/Fixity/Internal.hs:287]: Just look at this nonsense.

File: prettyprinter-configurable/src/Text/PrettyBy/Fixity.hs [hunk 4]
```diff
@@ -29,112 +29,122 @@ import Lens.Micro
...
 
--- | Instantiate a supplied continuation with a pretty-printer specialized to supplied
--- 'Fixity' and 'Direction'.
+{-| Instantiate a supplied continuation with a pretty-printer specialized to supplied
+'Fixity' and 'Direction'.
+-}
 withPrettyAt
-    :: MonadPrettyContext config env m
-    => Direction -> Fixity -> (AnyToDoc config ann -> m r) -> m r
+  :: (MonadPrettyContext config env m)
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/PrettyBy/Fixity.hs:100]: Is there any option to make it not add the redundant parens?

File: prettyprinter-configurable/src/Text/PrettyBy/Monad.hs [hunk 5]
```diff
@@ -5,40 +5,44 @@
...
 class HasPrettyConfig env config | env -> config where
-    prettyConfig :: Lens' env config
+  prettyConfig :: Lens' env config
 
 -- @env@ is an artefact of the encoding, it shouldn't be necessary as @m@ determines what it is
 -- and we're not interested in reflecting @env@ explicitly (unlike @config@, which is also
 -- determined by @m@ through @env@, but is useful to have explicitly). But GHC does not like it
 -- when @env@ is left implicit, see https://gitlab.haskell.org/ghc/ghc/issues/3490
+
 -- | A constraint for \"@m@ is a monad that allows to pretty-print values in it by a @config@\".
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/PrettyBy/Monad.hs:31]: Hate. The. Newline.

Like it at least could make it `--\n` instead to connect the comments.

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs [hunk 6]
```diff
@@ -574,256 +617,339 @@ application is implemented.
 
 -- | Throw @err@ when @b@ is stuck.
 type family ThrowOnStuck err (b :: Bool) :: Bool where
-    ThrowOnStuck _   'True  = 'True
-    ThrowOnStuck _   'False = 'False
-    ThrowOnStuck err _      = err
+  ThrowOnStuck _ 'True = 'True
+  ThrowOnStuck _ 'False = 'False
+  ThrowOnStuck err _ = err
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:622]: Huh, this won't conflict with `stylish-haskell`?

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs [hunk 7]
```diff
@@ -574,256 +617,339 @@ application is implemented.
...
+    TypeError
+      ( 'Text "No ’HasPrettyDefaults’ is specified for " ':<>: 'ShowType config
+          ':$$: 'Text "Either you're trying to derive an instance, in which case you have to use"
+          ':$$: 'Text "  standalone deriving and need to explicitly put a ‘PrettyDefaultBy config’"
+          ':$$: 'Text "  constraint in the instance context for each type in your data type"
+          ':$$: 'Text "  that supports default pretty-printing"
+          ':$$: 'Text "Or you're trying to pretty-print a value of a type supporting default"
+          ':$$: 'Text
+                  "  pretty-printing using a config, \
+                  \for which ‘HasPrettyDefaults’ is not specified."
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:639]: I hate `fourmolu`. This way the error message code was formatted directly reflected how the error message would end up being shown to the user. Now it's just random fucking garbage.

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs [hunk 8]
```diff
@@ -70,167 +70,170 @@ import GHC.TypeLits
...
-    -- Defaulting to 'prettyList' would require the user to either provide a 'Pretty' instance for
-    -- each data type implementing 'PrettyBy' or to supply the sensible default implementation as
-    -- 'defaultPrettyFunctorBy' manually and both the options are silly.
-    -- | 'prettyListBy' is used to define the default 'PrettyBy' instance for @[a]@ and @NonEmpty a@.
-    -- In normal circumstances only the 'prettyBy' function is used.
-    -- The default implementation of 'prettyListBy' is in terms of 'defaultPrettyFunctorBy'.
-    prettyListBy :: config -> [a] -> Doc ann
-    default prettyListBy :: config -> [a] -> Doc ann
-    prettyListBy = defaultPrettyFunctorBy
+  -- | Pretty-print a value of type @a@ the way a @config@ specifies it.
```
Comment (effectfully) [inline prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:108]: Stupid tool can't consistently replace `-- |` with `{-|`.

File: prettyprinter-configurable/src/Text/Fixity/Internal.hs (hunk 1 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/Fixity/Internal.hs:265]: @zliu41 agreed to `leading` if we do it fast :)

File: prettyprinter-configurable/src/Text/PrettyBy/Monad.hs (hunk 5 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/PrettyBy/Monad.hs:31]: One is internal comment, other is a public documentation. I want to be able to quickly and visually distinguish them. Newline is a nice way of separating them. Now I am not going to accidentally think this is a single logical description.

File: prettyprinter-configurable/src/Text/PrettyBy/Fixity.hs (hunk 4 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/PrettyBy/Fixity.hs:100]: Yes:
```yaml
# Whether to put parentheses around a single constraint (choices: auto, always, or never)
single-constraint-parens: always
```

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs (hunk 8 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:108]: Agree.

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs (hunk 6 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:622]: It probably does. My understanding is that we're going to stop using stylish haskell once we have adopted fourmolu.

File: prettyprinter-configurable/src/Text/PrettyBy/Internal.hs (hunk 7 above)
Comment (Unisay) [reply prettyprinter-configurable/src/Text/PrettyBy/Internal.hs:639]: This is not fourmolu, I wrapped this long line (>100) manually.

File: prettyprinter-configurable/src/Text/Fixity/Internal.hs [hunk 9]
```diff
@@ -260,19 +269,21 @@ encloseIn
...
+                  -- parens are needed in @x * (y + z)@.
+    EQ ->
+      case (assocOut, dir) of  -- If precedences are equal, then judge from associativity.
+        _ | assocOut /= assocInn       -> parens  -- Associativities differ => parens are needed.
+        (LeftAssociative, ToTheLeft)   -> id      -- No need for parens in @Add (Add x y) z@
+                                                  -- which is rendered as @x + y + z@.
+        (RightAssociative, ToTheRight) -> id      -- No need for parens in
+                                                  -- @Concat xs (Concat ys zs)@
+                                                  -- which is rendered as @xs ++ ys ++ zs@.
+        _                              -> parens  -- Every other case requires parens.
```
Comment (Unisay) [inline prettyprinter-configurable/src/Text/Fixity/Internal.hs:289]: Fourmolu disabled. 
The trade-off here: current formatting isn't diff-friendly (comment blocks shift horizontally depending on the lhs expression length) but is fitted to this particular code width thus making it more compact.

My preference is in making it less fragile by placing comments above the expressions with the same indentation level, such that it doesn't depend on the expression length:

```suggestion
  case precOut `compare` precInn of
    LT -> 
      -- If the outer precedence is lower than the inner, then
      -- do not add parens. E.g. in @Add x (Mul y z)@ the precedence
      -- of @Add@ is lower than the one of @Mul@, hence there is
      -- no need for parens in @x + y * z@.
      id
    GT ->
      -- If the outer precedence is greater than the inner, then
      -- do add parens. E.g. in @Mul x (Add y z)@ the precedence
      -- of @Mul@ is greater than the one of @Add@, hence
      -- parens are needed in @x * (y + z)@.
      parens
    EQ ->
      -- If precedences are equal, then judge from associativity.
      case (assocOut, dir) of 
        _ | assocOut /= assocInn ->
          -- Associativities differ => parens are needed.
          parens 
        (LeftAssociative, ToTheLeft) ->
          -- No need for parens in @Add (Add x y) z@
          -- which is rendered as @x + y + z@.
          id 
        (RightAssociative, ToTheRight) -> 
          -- No need for parens in
          -- @Concat xs (Concat ys zs)@
          -- which is rendered as @xs ++ ys ++ zs@.
          id
        _ ->
          -- Every other case requires parens.
          parens 
```

Such code trades vertical space (which I have plenty of) for precious horizontal space (which I like to save).

**Comment by effectfully** [comment]
> How about I instead resolve https://github.com/IntersectMBO/plutus/issues/4921 and we won't need to do anything here? This package hasn't been updated for a while anyway, so it's pretty stable.

So I did that instead: #6699. If that PR is merged, we'll close this one.

**Review by zliu41 — COMMENTED** [review]
Looks tolerable.

Besides leading arrows, what are the other non-default config options this uses?

**Comment by effectfully** [comment]
Closing in favor of #6699.

**Pull request closed without merging**